accepted/2021/InboxSourceGenerators.md

```diff
@@ -0,0 +1,191 @@
+4. ~~Automatically update user assembly though a hotpatch applied by runtime or JIT~~
+
+We will need to better understand what customers' expectations are around servicing to decide if we should do any work on these axes.  My initial inclination is to only have a PDB indicator and require users to subscribe to updates.
+
+### Functionality concerns for source generators
+
+Though the following are not specific to inbox source generators, these concerns are worth reviewing to describe how they apply to framework specific source generators.
+
+Source generator assemblies execute in the compiler process and must target .NET Standard in order run on both .NET Framework and .NET Core. Source generators should support generating source for all target frameworks supported by their companion library.  In the case of inbox source generators discussed in this document this need only be the target framework the generator ships in.  In the case of a NuGet package provided source generator, this should be all frameworks supported by package.
+
+Source generators should limit their dependencies.  There is no scheme for encoding framework-specific nor runtime-specific implementations of dependencies or source generators based on the runtime environment so source generators should avoid any dependencies on assemblies which need to differ by framework or runtime.  Source generators may depend on packages which are known dependencies of the compiler and should depend on versions less than or equal to that provided by the compiler and should not include that dependency in their package.
+
+Source generators should not rely on referencing nor executing framework types they wish to extend.  Despite inbox source generators being coupled to a specific framework version, source generators cannot assume they will be running on that framework.  They may be running on a newer, older, or completely different framework.  For example, API added to help a source generator do its job may not be present when running on .NET Framework or the .NET Core version the compiler is running on.  There is no type-equivalence guaranteed between types loaded in the compiler process and those of the user code that is under analysis.  For example, a source generator cannot instantiate attribute instances in the compiler process from user references to those attributes in source.  Should a source generator need to identify type usage it will need to examine user code references by name or resolve types in the user's references to ensure valid equivalence checks.  Source sharing should be considered as an alternative for cases where source generators need to execute runtime library helper code when generating source.  For example, serializer code which needs to examine the shape of user types can be shared between the runtime library and source generator.  Common abstractions can be made over metadata to allow the same code to handle different sources of metadata (source vs reflection).
```

Comment:
> Despite inbox source generators being coupled to a specific framework version, source generators cannot assume they will be running on that framework

Can we make this language stronger? This sounds like the normal case is they run on the same framework and that seeing other frameworks is an uncommon event. The opposite is the truth. 

> Should a source generator need to identify type usage it will need to examine user code references by name or resolve types in the user's references to ensure valid equivalence checks

Can we reference the source generators cookbook here? It has examples on how to do this correctly.